### PR TITLE
Add slack plugin

### PR DESCRIPTION
**Pull request by rthill**
New slack plugin for smarthomeNG

### CODE REVIEW

**Comment by rthill** [comment]
Yes, sure, I will update the plugin documentation according to the latest smarthomeNG release. I am using this plugin for almost two years and realized now that I didn't make it public.                                  Raoul                                                                                                                                                                           From: notifications@github.comSent: June 2, 2018 19:36To: plugins@noreply.github.comReply-to: reply@reply.github.comCc: raoul.thill@gmail.com; author@noreply.github.comSubject: Re: [smarthomeNG/plugins] Add slack plugin (#144)  @ohinckel commented on this pull request.

Could you also provide a plugin.yaml? The docs about how to write a plugin will help you.

**Comment by psilo909** [comment]
@rthill any update on the plugin.yaml? it would be pretty good to have the plugin based upon the technology of the upcoming (july) release. otherwise we will have to update all outdated plugins (i already did for about a dozen)

**Comment by msinn** [comment]
The parameter section of the plugin.yaml is incorrect. Please have a look at the documentation at https://www.smarthomeng.de/dev/developer/development_plugin/plugin_metadata.html

**Comment by maddinador** [comment]
Sorry for dumb question - my first time with git and github.
I've updated these files (__init__.py, plugin.yaml, README.md).
It seems that I do not have the rights to push my commit to pr/144. Shall I open a new pr with my changes? But then there is no connection to this pr any more.
Please give me a hint, thanks!

**Comment by ohinckel** [comment]
Usually others are not allowed to modify a user's branch. It is possible to allow other users to modify them by adding the users as collaborator to the repository. But I don't know if @rhill want this.

@maddinador, what are these changes? Has they something to do with this PR? Are they based on the changes from @rhill?

**Comment by rthill** [comment]
@maddinador you should have received a invitation to collaborate on the slack plugin and you should be able to update this PR.

**Comment by maddinador** [comment]
@rthill thanks, but that seems to be a separate repo without any connection to this pr. It even has different checkin dates and ids.
I've switched "plugins" to branch "pr/144" but I am not allowed to push anything there. Maybe this is not possible, I've only seen a option to allow maintainers to commit to a pr.

@ohinckel I've included the changes to plugin.yaml as requested by @msinn in this pr. Afterwards I've tested it on a fresh install. A Slack auth token is limited to a single channel, thus I've enabled multi-instance feature.

What is the better way: Shall I create a new pull request with a link to this one or shall I commit to the separate repo of rthill and he copies the files over and pushes them here? 😄

**Comment by rthill** [comment]
@maddinador okay, sorted out the thing. Now you should be able to collaborate on my plugin fork.

**Comment by maddinador** [comment]
@rthill: now it worked! 👍
Did you test posting to different slack channels some time ago? Were you able to post to different channels using the same token? API doc says that that should not be possible so I changed the plugin to multi-instance. Multi-Instance works here for posting to tweo different channels with different tokens.

Is someone able to review my changes?

**Comment by rthill** [comment]
Yes, posting to different Slack channels using the same token is possible and I am using it multiple times on daily basis and not only in smarthomeNG. I use different tokens only for different apps.

**Comment by maddinador** [comment]
oh, you're right. Did a short test for that, too.
I'll update the readme again.
But multi-instance can stay activated, can't it?

**Comment by rthill** [comment]
There might be users operating one smarthomeNG instance for office buildings and multiple sub-tenants which might make use of multi-instance, but besides that I cannot imagine a use case.
In my opinion it could be removed, as shNG is intended for home use.

**Comment by maddinador** [comment]
yeah, I understand that. Currently I don't see a use case, too. A Slack workspace is intended for a company and you usually don't work in several companies at a time.

From my point of view the code is stateless and only consists of a function "notify" with a webservice call in it. Thus I don't see any risks allowing multiple instances for future use.
Would there be any disadvantages if we let the plugin be multi-instance-able and one only uses a single instance? @ohinckel @msinn, what do you think?
(The change would me minimal: Only one line in init.py and plugin.yaml would need to be changed to enable/disable the multi-instace feature for this plugin.)

**Comment by ohinckel** [comment]
I don't know a lot about Slack. AFAIS, the difference between several multi instance configurations is the token parameter.

Different tokens are required for using different workspaces. Maybe there is an use case for this (posting to other workspaces for different messages, to separate them) so multi instance could be helpful here.

I don't have an opinion about to remove it keep it, but if we keep it the documentation should clarify why multi instance is useful or for what it can be used.

**Comment by ohinckel** [comment]
This PR is agains the `master` branch. Please create a PR against the `develop` branch.

**Comment by maddinador** [comment]
@rthill, your job! I am not allowed to. https://help.github.com/articles/changing-the-base-branch-of-a-pull-request/

File: slack/__init__.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        '<': "&lt;",
+        '"': "\"",
+        "'": "\'",
+    }
+
+    def __init__(self, smarthome, token):
+        self.logger = logging.getLogger(__name__)
+        self.logger.info("Init Slack notifications")
+        self._sh = smarthome
+        self.token = token
```
Comment (ohinckel) [inline slack/__init__.py]: Maybe you want to make use of `self.get_parameter_value('token')`.

File: slack/__init__.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    def run(self):
+        pass
+
+    def stop(self):
+        pass
+
+    def html_escape(self, text):
+        return "".join(self.html_escape_table.get(c, c) for c in text)
```
Comment (ohinckel) [inline slack/__init__.py]: Escaping stuff for HTML can also be done by using `html.escape()` method as documented [here](https://docs.python.org/3/library/html.html).

**Comment by ohinckel** [comment]
Thanks for taking care!

The ` __init__()` method looks OK to me. You don't have to list the parameters there anymore when using `get_parameter_value()`  to retrieve a parameter's value (great, this sound very logical).

What I meant with 
> Please create a PR against the develop branch.

... was, that this PR wants to be merged into `master` branch as indicated on top of this request:
> rthill wants to merge 6 commits into smarthomeNG:master from rthill:slack 

This should be easily changeable by just editing the PR and change the target branch to `develop` (or just create a new PR with correct branch).

**Comment by maddinador** [comment]
@ohinckel, Thanks for your review and your patience! This is my first engagement here and I am happy to learn!

I am not the owner of the PR and do not have the rights to change the branch. I think only @rthill could do this - or maybe a repo maintainer?
I'll wait some time for @rthill to change the branch, that'll be more “correct“ than creating a new PR. But I could do this to learn more!

**Comment by maddinador** [comment]
I've **tried** to sync rthill's fork to plugins:develop - now it seems like every file has a change "No change." 👎 
Maybe I've done something wrong following [https://help.github.com/articles/syncing-a-fork/](https://help.github.com/articles/syncing-a-fork/) in my commit a4308c0!? (only did a "checkout slack" to be in the right branch and a "merge upstream/develop" to be synced)
Shall I create a new fork/PR where I only add the new slack folder with three files agains develop?
Or is that current state ok for merging?

Travis-ci fails due to a test in `tests/test_logics.py` like other commits (only with py3.7)
Has anyone already checked that?

**Comment by ohinckel** [comment]
The difference for the file with "No changes." are only file mode changes (rwx).

> Or is that current state ok for merging?

I think a new PR is better. This PR contains much more changes (did not review all of them to find out what went wrong).

> Has anyone already checked that?

You created the issue https://github.com/smarthomeNG/smarthome/issues/294 already.

**Comment by ohinckel** [comment]
Closing this, new PR is #174.

**Pull request closed without merging**